### PR TITLE
Fix ensembles to match Lopes de santos matlab code

### PR DESCRIPTION
**Pull request by neuromantic99**
## Summary by CodeRabbit

- New Features
  - ICA-based ensemble reactivation analysis with expanded plots and diagnostics.
  - Optional processing for sessions lacking wheel-freeze data.
  - New session summary output exposing ensemble metrics.
  - New utilities to split continuous data chunks and detect threshold crossings.

- Improvements
  - More robust metadata handling and NaN-aware statistics.
  - Natural sorting of trial files and updated default cache locations.
  - Updated debug launch configurations to adjust slow-resolve warnings.
  - Broad dependency reconfiguration impacting runtime environments.

- Tests
  - Added unit tests for new utilities and concurrent shuffling behavior.

- Documentation & Chores
  - Updated setup instructions; added nohup.out to .gitignore.

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -549,3 +552,45 @@ def test_compute_windowed_speed_movement_and_at_the_edges() -> None:
 #     ax2.set_xlabel("Time (s)")
```
Comment (josefbitzenhofer) [inline tests/test_utils.py:544]: I know this is old code, but do we want to keep it as it is commented-out?

File: viral/run_oasis.py [hunk 2]
```diff
@@ -333,13 +341,31 @@ def main(
...
+            print(
+                f"No wheel freeze info for {mouse_name} {date}, haven't dealt with this in oasis yet so skipping"
+            )
+            continue
+
+        try:
+            main(s2p_path, cached_session.wheel_freeze, parallel=False, plot=False)
+        except Exception as e:
+            print(f"Error processing {mouse_name} {date}: {e}")
+            continue
```
Comment (josefbitzenhofer) [inline viral/run_oasis.py:371]: All good but as you mentioned we should deal with no-wheel-freeze sessions and perhaps we could offer an easy way for single session oasis processing as well

File: viral/imaging_utils.py [hunk 3]
```diff
@@ -197,6 +192,7 @@ def get_online_position_and_frames(
     )
 
     # Removed the first two seconds as there is a bit of a burst of activity when the screens come on, which is not unexpected
+    # TODO: Not sure this is working correctly
     trial_onset = frame_position < frame_position[0] + 60
     # TODO: Make sure this works
```
Comment (josefbitzenhofer) [inline viral/imaging_utils.py:197]: Well... are you sure now? 👉🏻👈🏻

File: viral/ensemble_reactivation.py [hunk 4]
```diff
@@ -96,32 +114,90 @@ def get_running_bouts(
...
+        whiten="unit-variance",
+        fun="logcosh",  # logcosh with alpha = 1 is the same as tanh used in matlab
+        fun_args={"alpha": 1.0},
+        max_iter=500,
+        random_state=0,  # Optional: for reproducibility
+    )
+    S = ica.fit_transform(X_centered)  # Shape: (n_timepoints, n_components)
+    W = ica.components_  # Shape: (n_components, n_cells)
+
+    return W.T
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:196]: We tested it's the same, right?

File: viral/ensemble_reactivation.py [hunk 5]
```diff
@@ -337,6 +411,7 @@ def sort_ensembles_by_reactivation_strength(
     # total_strength = np.sum(positive_only, axis=1)
 
     total_strength = np.sum(reactivation_strength, axis=1)
+    # total_strength = np.max(zscore(reactivation_strength, axis=1), axis=1)
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:413]: I assume the not-commented-out one is correct? In the diff, the only thing that changed was the the commented-out version being added

File: viral/ensemble_reactivation.py [hunk 6]
```diff
@@ -752,17 +848,19 @@ def main() -> None:
             # running_bouts=,
             pcc_scores=pcc_scores,
         )
+        # if not plot:
+        #     return
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py]: currently the plot argument defaults to True but is never accessed in the code function -> should we delete that argument? i mean later on we might need this function to return something but at the moment the return is None as well

Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:852]: there is an unused `plot` argument in this function, should we perhaps get rid of it for now as the function doesn't return anything at the moment and plotting and saving cache is the only functionality

File: viral/ensemble_reactivation.py [hunk 7]
```diff
@@ -579,16 +655,61 @@ def plot_grosmark_panel(
...
+            ssp_vectors.append(
+                np.apply_along_axis(
+                    gaussian_filter1d,
+                    axis=1,
+                    arr=place_cells[:, chunk],
+                    sigma=sigma,
+                )
+            )
+
+    return np.hstack(ssp_vectors)
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:715]: I get this is online activity w/o ITI and ssp vectors are described in Grosmark, but perhaps we could have a one liner docstring here? 🙈

File: viral/ensemble_reactivation.py [hunk 8]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
         ax1.legend(lines, labels, loc="upper left")
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py]: Has later come now and we want to delete this?

File: viral/ensemble_reactivation.py [hunk 9]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
...
+def compare_to_matlab() -> None:
+
+    # test_data = np.load(
+    #     "/Volumes/hard_drive/VR-2p/2025-07-05/JB036/suite2p/plane0/oasis_spikes.npy"
+    # )
+    # test_data = gaussian_filter1d(test_data, sigma=30, axis=1)
+
+    # # take a random-ish subset of the online data
+    # test_data = test_data[:, 27000:100000]
+    # np.save("test_ensemble_data.npy", test_data)
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py]: perhaps we could delete the commented-out lines but keep the rest in this function?

File: viral/ensemble_reactivation.py [hunk 10]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
...
+
+    all_mice: List[EnsembleSessionResult] = []
+
+    baseline_multiplier = 1
+    for cache_file in cache_files:
+
+        mouse, date = cache_file.stem.split("_")[:2]
+        mouse = mouse.strip("suite2p")  # dunno why this is in the path lol
+        if mouse not in {"JB030", "JB031"}:
+            continue
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py]: do you want to keep filtering for certain mice? do we want to have a list of mouse names as an argument to this function?

File: viral/ensemble_reactivation.py [hunk 11]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
...
+    )
+    shaded_line_plot(
+        response_pre,
+        x_axis=np.arange(-30, 30) / 30,
+        color="orange",
+        label="preactivation",
+    )
+    plt.xlabel("Time (s) from event onset")
+    1 / 0
+
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:1297]: whoopsie 
```suggestion

```

File: viral/ensemble_reactivation.py [hunk 12]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
...
+        response_re, x_axis=np.arange(-30, 30) / 30, color="blue", label="reactivation"
+    )
+    shaded_line_plot(
+        response_pre,
+        x_axis=np.arange(-30, 30) / 30,
+        color="orange",
+        label="preactivation",
+    )
+    plt.xlabel("Time (s) from event onset")
+    1 / 0
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/ensemble_reactivation.py]: whopsie
```suggestion

```

File: viral/ensemble_reactivation.py [hunk 13]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
...
+        )
+    all_mouse_plots(all_mice=all_mice)
+
+
+def all_mouse_plots(
+    all_mice: List[EnsembleSessionResult],
+) -> None:
+    number_of_events = np.hstack(
+        [mouse.number_of_events[0] - mouse.number_of_events[1] for mouse in all_mice]
+    )
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:1231]: do you still need this variable?

File: viral/ensemble_reactivation.py [hunk 14]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
...
+            EnsembleSessionResult(
+                reactivation_triggered_response=reactivation_triggered_response,
+                number_of_events=reactivation_number_of_events,
+                sum_values_over_threshold=reactivation_strength_sum_over_threshold,
+            )
+        )
+    all_mouse_plots(all_mice=all_mice)
+
+
+def all_mouse_plots(
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py]: I know this is really picky but perhaps we could rename it something more descriptive like `all_mice_ensemble_results_plots` even though it is longer? It's obvious that it's ensemble reactivation as this is the file name but it's the summary of the ensemble reactivation as I understood.

File: viral/ensemble_reactivation.py [hunk 15]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
...
+
+    test_data = np.load("test_ensemble_data.npy")
+
+    python_result = compute_ICA_components(test_data)
+
+    compare_run_results(matlab_result, python_result)
+
+
+def multiple_sessions() -> None:
+
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:1142]: Perhaps a short docstring or description?

File: viral/ensemble_reactivation.py [hunk 16]
```diff
@@ -790,27 +888,167 @@ def main() -> None:
...
+        plt.ylabel("Reactivation strength (sum across time)")
+        plt.title(
+            f" Sum of values over threshold: Mean change = {np.mean(total_reactivation - total_preactivation):.2f} p ={wilcoxon(total_reactivation, total_preactivation)[1]:.2f}",
+        )
+    return total_reactivation, total_preactivation
+
+
+def reactivation_triggered_average(
+    data: np.ndarray, baseline: np.ndarray, length_event_samples: int = 30
+) -> np.ndarray:
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:941]: Perhaps a short docstring?

File: viral/ensemble_reactivation.py [hunk 17]
```diff
@@ -862,5 +1100,214 @@ def plot_speed_and_position_logic(trials: List[TrialInfo]) -> None:
...
+        label="preactivation",
+    )
+    plt.xlabel("Time (s) from event onset")
+    1 / 0
+
+
 if __name__ == "__main__":
-    main()
+    # This is the file that's saved by the full grosmark oasis preprocessing as a flag
+    valid_sessions = list(TIFF_UMBRELLA.rglob("full_grosmark_oasis_preprocessed.npy"))
```
Comment (josefbitzenhofer) [inline viral/ensemble_reactivation.py:1264]: This flag is super important and I think this is helpful for all kinds of analyses (I'll retry my sequences with the "good"/"valid" grosmark oasis). Hence, I think we should think about using this flag in a util

**Review by josefbitzenhofer — COMMENTED** [review]
See comments

File: viral/ensemble_reactivation.py [hunk 18]
```diff
@@ -579,16 +655,61 @@ def plot_grosmark_panel(
...
+            if len(chunk) < 2 * 30:  # Arbitrary removal of short chunks
+                continue
+            ssp_vectors.append(
+                np.apply_along_axis(
+                    gaussian_filter1d,
+                    axis=1,
+                    arr=place_cells[:, chunk],
+                    sigma=sigma,
+                )
+            )
```
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/ensemble_reactivation.py]: but place_cells[:, chunk] is 2D?

File: viral/ensemble_reactivation.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/ensemble_reactivation.py]: yeahhhhh ok this shouldn't be commited, but I kinda like 1 / 0 at the end of lots of plots when you aren't gonna run anything further. it's better than plt.show() becuase you still have access to all the variables in the debug console and it opens all the plots

[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/ensemble_reactivation.py]: what do you think coderabbit i value your opinion

File: viral/utils.py [hunk 19]
```diff
@@ -142,6 +142,39 @@ def threshold_detect(signal: np.ndarray, threshold: float) -> np.ndarray:
...
+
+    # Compare elementwise
+    thresh_signal = signal > threshold
+
+    # Keep only rising edge detections
+    thresh_signal[1:][thresh_signal[:-1] & thresh_signal[1:]] = False
+
+    # Return indices
+    times = np.where(thresh_signal)
+    return times[0]
```
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/utils.py:175]: if your analysis is correct why does this test pass?

```python
def test_threshold_detect_continuous() -> None:
    arr = np.array([0, 0, 0, 10, 10, 10, 0])
    threshold = np.array([1, 1, -10, 9, 9, 100, 0])

    result = threshold_detect_continuous(arr, threshold)
    expected = np.array([2])
    assert np.array_equal(result, expected)
    ```

File: tests/test_utils.py (hunk 1 above)
Comment (neuromantic99) [reply tests/test_utils.py:544]: yeah could be deleted, but it's kind of useful and should be commented out as in the docstring:

"""Don't have this as part of the actual test suite as it requires server access"""

File: viral/run_oasis.py (hunk 2 above)
Comment (neuromantic99) [reply viral/run_oasis.py:371]: Already fixed in learning-stage-improvments :)

File: viral/imaging_utils.py (hunk 3 above)
Comment (neuromantic99) [reply viral/imaging_utils.py:197]: yeahhh i think the literal code works but i'm not sure yet if it's a good idea to do it, i'll add a ticket for it

File: viral/ensemble_reactivation.py (hunk 4 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py:196]: i did a while ago yes, you can double check with compare_run_results if you like

File: viral/ensemble_reactivation.py (hunk 5 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py:413]: dunno the strength sorting might not even end up being used, can look into it closely if we decide to

File: viral/ensemble_reactivation.py (hunk 6 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py]: oh yeah i used that to loop through sessions and save them on the server, i'll add it back in

File: viral/ensemble_reactivation.py (hunk 7 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py:715]: nah self documenting code mate ;)

Comment (josefbitzenhofer) [reply viral/ensemble_reactivation.py:715]: Whatever you say

File: viral/ensemble_reactivation.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/ensemble_reactivation.py]: Re-summoning @coderabbitai here

[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/ensemble_reactivation.py]: nah lame

File: viral/ensemble_reactivation.py (hunk 8 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py]: haha i don't know what this does anymore so i guess so

File: viral/ensemble_reactivation.py (hunk 9 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py]: i actually think the whole thing should be deleted, the juicy hard stuff is in compare_run_results which we'll keep. This bascially just loads data that I don't even know the location of anymore

File: viral/ensemble_reactivation.py (hunk 10 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py]: yeah this is dumb, have removed :)

File: viral/ensemble_reactivation.py (hunk 13 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py:1231]: maybe, it's not used currently but it may be a metric we want to use down the line

File: viral/ensemble_reactivation.py (hunk 17 above)
Comment (neuromantic99) [reply viral/ensemble_reactivation.py:1264]: how would you use the flag in a util?
